sys/arm64/vmm/arm64.h

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) 2015 Mihai Carabas <mihai.carabas@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#ifndef _VMM_ARM64_H_
+#define _VMM_ARM64_H_
+
+#include <machine/reg.h>
+#include <machine/hypervisor.h>
+#include <machine/pcpu.h>
+
+#include "mmu.h"
+#include "io/vgic_v3.h"
+#include "io/vtimer.h"
+
+struct vgic_v3;
+struct vgic_v3_cpu;
+
+struct hypctx {
+	struct trapframe tf;
+
+#if __has_feature(capabilities)
+	uint64_t	cctlr_el0;
+	uint64_t	cctlr_el1;
+	uintcap_t	cid_el0;
+	uintcap_t	ddc_el0;
+	uintcap_t	rcsp_el0;
+	uintcap_t	rddc_el0;
+	uintcap_t	rctpidr_el0;
+#endif
+
+	/*
+	 * EL1 control registers.
+	 */
+#if __has_feature(capabilities)
+	uintcap_t	elr_el1;	/* Exception Link Register */
+	uintcap_t	sp_el0;		/* Stack pointer */
+	uintcap_t	tpidr_el0;	/* EL0 Software ID Register */
+	uintcap_t	tpidrro_el0;	/* Read-only Thread ID Register */
+	uintcap_t	tpidr_el1;	/* EL1 Software ID Register */
+	uintcap_t	vbar_el1;	/* Vector Base Address Register */
+#else
+	uint64_t	elr_el1;	/* Exception Link Register */
+	uint64_t	sp_el0;		/* Stack pointer */
+	uint64_t	tpidr_el0;	/* EL0 Software ID Register */
+	uint64_t	tpidrro_el0;	/* Read-only Thread ID Register */
+	uint64_t	tpidr_el1;	/* EL1 Software ID Register */
+	uint64_t	vbar_el1;	/* Vector Base Address Register */
```

Comment:
Just use uintcap_t? We're not doing ILP32 AArch64 in FreeBSD. We just use uintcap_t for trapframe.

Reply:
Sorry, I'm not sure what you mean here. We use uintcap_t for these fields when `__has_feature(capabilities)` is true.

Reply:
uintcap_t is uintptr_t which is uint64_t in the `! __has_feature(capabilities)` case so can be used unconditionally here.

Reply:
Right, but the definition of `struct trapframe` uses an ifdef the same way I did here, and it seemed sensible to copy that approach. The code which populates this structure needs an ifdef since `READ_SPECIALREG_CAP` is defined only when `__has_feature(capabilities)`, so it also felt natural to duplicate the field definitions.

I can change it of course, but I don't understand "We just use uintcap_t for trapframe."